Adding the Datadog monitor API

onlyanegg opened the pull request:
Extending the Datadog execution module to include the monitor API.
Adding a Datadog state module including monitor API functions.
Adding a Datadog utils module for common functions.

### What does this PR do?
This PR adds executions module functions and state module functions for the datadog monitor api

### What issues does this PR fix or reference?

### Tests written?

No

### Commits signed with GPG?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

onlyanegg committed:
Adding the Datadog monitor API

Extending the Datadog execution module to include the monitor API.
Adding a Datadog state module including monitor API functions.
Adding a Datadog utils module for common functions.

rallytime commented on salt/modules/datadog_api.py:
```diff
@@ -262,3 +259,426 @@ def post_event(api_key=None,
+                   query=None,
+                   name=None,
+                   message=None,
+                   tags=[],
```
I see your comment above, but we need to only use `None` as defaults here. Then you can handle creating lists within the function if `None`.

rallytime commented on salt/modules/datadog_api.py:
```diff
@@ -262,3 +259,426 @@ def post_event(api_key=None,
+                 id=None,
+                 name=None,
+                 group_states=None,
+                 tags=[],
```
Same comment for these kwargs as above.

rallytime commented on salt/modules/datadog_api.py:
```diff
@@ -262,3 +259,426 @@ def post_event(api_key=None,
+                   id=None,
+                   name=None,
+                   query=None,
+                   message='',
```
Same comment here. :)

Instead of me commenting on all of the functions that do this, the functions in the state below that behave this way should also be addressed.

rallytime reviewed (changes requested):
Hi again @onlyanegg - good to see you. Thank you for adding this additional functionality! I have a couple of comments on this that need to be addressed. Can you take a look?

onlyanegg commented:
Hi, @rallytime. Thanks for taking a look! I'll definitely address those concerns.

onlyanegg commented:
Hi, @rallytime. I stumbled across some other issues while dealing with the argument default values including checking for `test=true` in the states and some flaky and incorrect code in the get defaults function. I changed the title to WIP until those can be addressed.

rallytime commented:
Sounds great. Thanks @onlyanegg!

onlyanegg committed:
Fixes for feedback and bugs

Changing the default arguments
Fix for composite monitor with ids
Adding test=true check for states

rallytime commented:
Hey @onlyanegg - just a reminder bump here to see if you had a chance to come back to this. This also has a merge-conflict here.

onlyanegg commented:
Hi, @rallytime. I apologize. I was very busy for more than a month, but I have had some time recently and do have some time currently, so I will be working on it. Currently, I'm writing tests to make sure the behavior is correct.

rallytime commented:
@onlyanegg Wonderful! Thank you so much.

To fix the merge conflict, you'll need to rebase your changes with the current HEAD of develop. Then you can fix the conflict locally. Note that when you push up your latest changes (because of the rebase) to GitHub you'll get a warning so you'll have to force push. Let me know if you have any other questions! :)

onlyanegg committed:
Bug fixes

- Add special case for updates to tags and monitor tags
- Using six.string_types in place of str for isinstance test to account for
  Python 2 str and unicode

onlyanegg committed:
Changing all message='' defaults to message=None

onlyanegg commented:
Hi, @rallytime! This is ready now. I do have tests for the execution module, but it requires a Datadog account. Would that still go in the salt repo?

garethgreenaway commented:
@onlyanegg Is there a way to run the tests using mocking to avoid the requirement of the Datadog account?

onlyanegg commented:
Hi, @garethgreenaway. I can't think of a way to effectively mock the datadog api calls. I'm checking the response to make sure that the operation was performed correctly, so returning a dummy response from a mock object seems to defeat the point. But I'm open to suggestions! I'm also using a free datadog account in the tests. I'm not sure if that makes a difference...

onlyanegg commented:
@rallytime, @garethgreenaway, is there anything else I should do for this to be merged?

Commit:
Merge branch 'develop' into datadog-monitor_api

garethgreenaway committed:
Merge branch 'develop' into datadog-monitor_api

garethgreenaway commented:
@onlyanegg Sorry for the delay, looks like there are some lint issues in the PR.

onlyanegg commented:
np, @garethgreenaway. I'll take a look.

garethgreenaway commented:
Awesome!  Thanks!

garethgreenaway committed:
Merge branch 'develop' into datadog-monitor_api

onlyanegg commented:
Hi, @garethgreenaway. I haven't had a chance to look at this in a while, but I hope to have some time soon.

rallytime commented:
Since we haven't heard back here for a while, I'm going to close this PR. Once those lint issues are addressed we will be happy to open this again, of course!

The jobs keep timing out, so I will just post the lint builds here:
```
➜  salt git:(pr-44799) pylint --rcfile=.testing.pylintrc  salt/states/datadog_api.py
Using config file /root/SaltStack/salt/.testing.pylintrc
/usr/local/lib/python3.5/dist-packages/saltpylint/minpyver.py:31: RuntimeWarning: Unable to import saltpylint.ext.pyqver. Running on Python 3? Checker skipped.
  RuntimeWarning
************* Module salt.states.datadog_api
salt/states/datadog_api.py:327: [W1699(incompatible-py3-code), ] Incompatible Python 3 code found. Proposed fix:
                     # show options that are not specified by the user.
                     pass
         elif attribute == 'tags':
-            current['tags'] = map(lambda s: s.strip(), current['tags'])
-            desired['tags'] = map(lambda s: s.strip(), desired['tags'])
+            current['tags'] = [s.strip() for s in current['tags']]
+            desired['tags'] = [s.strip() for s in desired['tags']]
         else:
             current[attribute] = current[attribute].strip()
             desired[attribute] = desired[attribute].strip()
salt/states/datadog_api.py:384: [C0303(trailing-whitespace), ] Trailing whitespace

-----------------------------------
Your code has been rated at 9.88/10

➜  salt git:(pr-44799) pylint --rcfile=.testing.pylintrc  salt/utils/datadog_api.py
Using config file /root/SaltStack/salt/.testing.pylintrc
/usr/local/lib/python3.5/dist-packages/saltpylint/minpyver.py:31: RuntimeWarning: Unable to import saltpylint.ext.pyqver. Running on Python 3? Checker skipped.
  RuntimeWarning
************* Module salt.utils.datadog_api
salt/utils/datadog_api.py:70: [E1133(not-an-iterable), find_monitors_with_name] Non-iterable value res['response'] is used in an iterating context

-----------------------------------
Your code has been rated at 8.89/10
```

Please let us know when this is ready to go and we will gladly open this again.
